x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/AbstractScriptFieldType.java

```diff
@@ -12,52 +12,58 @@
 import org.apache.lucene.search.spans.SpanQuery;
 import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.CheckedBiConsumer;
 import org.elasticsearch.common.TriFunction;
 import org.elasticsearch.common.geo.ShapeRelation;
 import org.elasticsearch.common.time.DateMathParser;
 import org.elasticsearch.common.unit.Fuzziness;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.mapper.ContentPath;
 import org.elasticsearch.index.mapper.DocValueFetcher;
+import org.elasticsearch.index.mapper.FieldMapper;
 import org.elasticsearch.index.mapper.MappedFieldType;
-import org.elasticsearch.index.mapper.TextSearchInfo;
+import org.elasticsearch.index.mapper.Mapper;
+import org.elasticsearch.index.mapper.MapperParsingException;
+import org.elasticsearch.index.mapper.RuntimeFieldType;
 import org.elasticsearch.index.mapper.ValueFetcher;
 import org.elasticsearch.index.query.QueryShardContext;
 import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptType;
 import org.elasticsearch.search.lookup.SearchLookup;
 
+import java.io.IOException;
 import java.time.ZoneId;
+import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
 
 import static org.elasticsearch.search.SearchService.ALLOW_EXPENSIVE_QUERIES;
 
 /**
  * Abstract base {@linkplain MappedFieldType} for scripted fields.
  */
-abstract class AbstractScriptFieldType<LeafFactory> extends MappedFieldType {
+abstract class AbstractScriptFieldType<LeafFactory> extends RuntimeFieldType {
     protected final Script script;
     private final TriFunction<String, Map<String, Object>, SearchLookup, LeafFactory> factory;
+    private final CheckedBiConsumer<XContentBuilder, Boolean, IOException> toXContent;
+
+    AbstractScriptFieldType(String name, TriFunction<String, Map<String, Object>, SearchLookup, LeafFactory> factory, Builder builder) {
+        this(name, factory, builder.script.getValue(), builder.meta.getValue(), builder::toXContent);
+    }
 
     AbstractScriptFieldType(
         String name,
-        Script script,
         TriFunction<String, Map<String, Object>, SearchLookup, LeafFactory> factory,
-        Map<String, String> meta
+        Script script,
+        Map<String, String> meta,
+        CheckedBiConsumer<XContentBuilder, Boolean, IOException> toXContent
     ) {
-        super(name, false, false, false, TextSearchInfo.SIMPLE_MATCH_WITHOUT_TERMS, meta);
-        this.script = script;
+        super(name, meta);
         this.factory = factory;
-    }
-
-    protected abstract String runtimeType();
-
-    @Override
-    public final String typeName() {
-        return RuntimeFieldMapper.CONTENT_TYPE;
-    }
-
-    @Override
-    public final String familyTypeName() {
-        return runtimeType();
+        this.script = script;
+        this.toXContent = toXContent;
```

Comment:
so, this is slightly hacky, it is done this way so that we get the Parameter serialization for free. We reuse the FieldMapper parsing code without having a FieldMapper. The constructors that take a function are there only for testing, because I don't want to have to create a FieldMapper.Builder in our tests.